File: .agent/skills/cnma-cap-backend/templates/java/srv/src/main/java/cnma/{{module_name}}/common/service/BaseService.java
package cnma.{{module_name}}.common.service;

import cnma.{{module_name}}.common.interfaces.ICommonService;
import cnma.{{module_name}}.common.model.ServiceResponse;
import cnma.{{module_name}}.common.enums.HttpStatusCode;
import lombok.extern.slf4j.Slf4j;

/**
 * Base service class providing common functionality.
 * EXTEND THIS CLASS for any new service that needs:
 * - Localized messages (getText)
 * - Standardized response formatting
 * - Audit fields (setCreateManaged, setUpdateManaged)
 */
@Slf4j
public abstract class BaseService implements ICommonService {

    @Override
    public String getText(String textKey) {
        return textKey;
    }

    @Override
    public String getText(String textKey, String... params) {
        String text = textKey;
        for (int i = 0; i < params.length; i++) {
            text = text.replace("{" + i + "}", params[i]);
        }
        return text;
    }

    protected <T> ServiceResponse<T> buildResponse(HttpStatusCode status, String message, T data) {
        return ServiceResponse.<T>builder()
                .statusCode(status.getCode())
                .success(status.isSuccess())
                .message(message)
                .data(data)
                .build();
    }

    protected <T> ServiceResponse<T> ok(T data) {
        return buildResponse(HttpStatusCode.OK, "Success", data);
    }

    protected <T> ServiceResponse<T> created(T data) {
        return buildResponse(HttpStatusCode.CREATED, "Created", data);
    }

    protected <T> ServiceResponse<T> error(HttpStatusCode status, String message) {
        return buildResponse(status, message, null);
    }
}

File: .agent/skills/cnma-cap-backend/templates/node/srv/src/common/service/BaseService.ts
import { ICommonService, IServiceResponse, IValidationResponseData, IApiResponse } from '../interfaces/ICommon';
import { HTTP_SUCCESS_CODES } from '../enum/HttpStatusCodeEnum';

/**
 * BaseService - Abstract base class for all services.
 * Provides: i18n text retrieval, response builders, managed field setters.
 *
 * EXTEND THIS CLASS for any new service that needs:
 * - Localized messages (getText)
 * - Standardized response formatting (buildServiceResponse, buildApiResponse)
 * - Audit fields (setCreateManaged, setUpdateManaged)
 */
export abstract class BaseService implements ICommonService {
    /**
     * Get localized text by key.
     * Override in subclass to load from CDS i18n.
     */
    getText(textKey: string, params?: string[]): string {
        let text = textKey;
        if (params && params.length > 0) {
            params.forEach((param, index) => {
                text = text.replace(new RegExp(`\\{${index}\\}`, 'g'), param);
            });
        }
        return text;
    }

    /**
     * Build service response with status code.
     */
    buildServiceResponse<T>(statusCode: number, message: string, data?: T): IServiceResponse<T> {
        const success = HTTP_SUCCESS_CODES.includes(statusCode);
        return {
            statusCode,
            success,
            message,
            data: data || null,
            toApiResponse(): IApiResponse<T> {
                return { success: this.success, message: this.message, data: this.data };
            }
        };
    }

    /**
     * Build validation response.
     */
    buildValidationResponse<T>(valid: boolean, message: string): IValidationResponseData<T> {
        return {
            valid,
            message,
            toApiResponse(): IApiResponse<T> {
                return { success: this.valid, message: this.message, data: null };
            }
        };
    }

    /**
     * Set create managed fields (audit trail).
     */
    setCreateManaged(data: any, owner: string): void {
        data.createdAt = new Date().toISOString();
        data.createdBy = owner;
        data.modifiedAt = new Date().toISOString();
        data.modifiedBy = owner;
    }

    /**
     * Set update managed fields.
     */
    setUpdateManaged(data: any, owner: string): void {
        data.modifiedAt = new Date().toISOString();
        data.modifiedBy = owner;
    }
}